### ISSUE TITLE
lodash Dependency

### ISSUE THREAD

**Issue by mxro** [cheeriojs/cheerio#1085]
I am interested in making cheerio as small as possible in terms of the footprint of its download size. There seems to be a [pull request](https://github.com/cheeriojs/cheerio/pull/864) which specifies individual lodash module dependencies rather than all of lodash.

Using the lodash dependency declarations from this package, I was able to reduce the download size of cheerio significantly (by a few mbs, see the [test project here](https://github.com/mxro/cheerio-lightweight)).

```
 "lodash.bind": "^4.1.4",
 "lodash.defaults": "^4.0.1",
 "lodash.filter": "^4.4.0",
 "lodash.flatten": "^4.2.0",
 "lodash.foreach": "^4.3.0",
 "lodash.map": "^4.4.0",
 "lodash.merge": "^4.4.0",
 "lodash.pick": "^4.2.1",
 "lodash.reduce": "^4.4.0",
 "lodash.reject": "^4.4.0",
 "lodash.some": "^4.6.0",
 "lodash.assignin" : "^4.2.0"
```

Looking at the [current version of package.json](https://github.com/cheeriojs/cheerio/blob/0f3d2dd9645b49dc4a54ddb14d6dd66cdfe2f32d/package.json) on the master branch, the lodash dependency declarations are reset to the complete lodash rather than the individual modules. Is there a reason why the individual module dependencies are not used?

(Please also see this related [Stackoverflow question](https://stackoverflow.com/questions/46335662/cheerio-js-lodash-dependency/46336410))

**Comment by billyjanitsch** [comment]
See the comments on #864, and #913.

tl;dr this would make cheerio smaller on its own, but would prevent deduping in the common case of having other dependencies which depend on Lodash. Also, it would load far more code at runtime.

**Comment by mxro** [comment]
Thanks for your response. That's good to know!

One thing I didn't fully understand was this "it would load far more code at runtime". Could you explain this a little further?

**Comment by billyjanitsch** [comment]
Sure. Many Lodash functions depend on each other. The zero-dependency modules (one per function) are compiled to inline their dependencies into a single file. Therefore, modules for functions A and B which both depend on C each contain an identical copy of C. If you depend on `lodash.A` and `lodash.B`, you load C twice. If you depend on Lodash and load `lodash/A` and `lodash/B`, you only load C once.

In practice, the dependency chains are very long. For instance, the source for [lodash.map](https://unpkg.com/lodash.map@4.6.0/index.js) is over 2000 lines long, as is the source for [lodash.filter](https://unpkg.com/lodash.filter@4.6.0/index.js), with most code being duplicated between them. Depending on 12 of them would mean loading something like ~10x as much Lodash code at runtime.

**Comment by fb55** [comment]
Awesome response @billyjanitsch
